<div>
<img src=https://www.institutedata.com/wp-content/uploads/2019/10/iod_h_tp_primary_c.svg width="300">
</div>

# Lab 4.1.2 Linear Regression

## Prediction of House Price Using Linear Regression

### Data

The Sacramento real estate transactions file is a list of 985 real estate transactions in the Sacramento area reported over a five-day period, as reported by the Sacramento Bee.

In [2]:
import pandas as pd
import numpy as np
import pandas as pd

%matplotlib inline

### 1. Explore Data

Perform EDA on this dataset. Find Dependent and Independent Variables to predict house price.

In [3]:
# Read CSV
house_csv = "C:/Users/jadea/Documents/IOD/04.1 - Module 04 Lab/Sacramento_transactions.csv"

In [5]:
housedata = pd.read_csv(house_csv)

In [6]:
housedata.head()

street        city    zip state  beds  baths  sq__ft  \
0      3526 HIGH ST  SACRAMENTO  95838    CA     2      1     836   
1       51 OMAHA CT  SACRAMENTO  95823    CA     3      1    1167   
2    2796 BRANCH ST  SACRAMENTO  95815    CA     2      1     796   
3  2805 JANETTE WAY  SACRAMENTO  95815    CA     2      1     852   
4   6001 MCMAHON DR  SACRAMENTO  95824    CA     2      1     797   

          type                     sale_date  price   latitude   longitude  
0  Residential  Wed May 21 00:00:00 EDT 2008  59222  38.631913 -121.434879  
1  Residential  Wed May 21 00:00:00 EDT 2008  68212  38.478902 -121.431028  
2  Residential  Wed May 21 00:00:00 EDT 2008  68880  38.618305 -121.443839  
3  Residential  Wed May 21 00:00:00 EDT 2008  69307  38.616835 -121.439146  
4  Residential  Wed May 21 00:00:00 EDT 2008  81900  38.519470 -121.435768

In [7]:
housedata.tail()

street             city    zip state  beds  baths  sq__ft  \
980   9169 GARLINGTON CT       SACRAMENTO  95829    CA     4      3    2280   
981      6932 RUSKUT WAY       SACRAMENTO  95823    CA     3      2    1477   
982    7933 DAFFODIL WAY   CITRUS HEIGHTS  95610    CA     3      2    1216   
983     8304 RED FOX WAY        ELK GROVE  95758    CA     4      2    1685   
984  3882 YELLOWSTONE LN  EL DORADO HILLS  95762    CA     3      2    1362   

            type                     sale_date   price   latitude   longitude  
980  Residential  Thu May 15 00:00:00 EDT 2008  232425  38.457679 -121.359620  
981  Residential  Thu May 15 00:00:00 EDT 2008  234000  38.499893 -121.458890  
982  Residential  Thu May 15 00:00:00 EDT 2008  235000  38.708824 -121.256803  
983  Residential  Thu May 15 00:00:00 EDT 2008  235301  38.417000 -121.397424  
984  Residential  Thu May 15 00:00:00 EDT 2008  235738  38.655245 -121.075915

In [8]:
housedata.isnull().sum()

street       0
city         0
zip          0
state        0
beds         0
baths        0
sq__ft       0
type         0
sale_date    0
price        0
latitude     0
longitude    0
dtype: int64

In [9]:
housedata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 985 entries, 0 to 984
Data columns (total 12 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   street     985 non-null    object 
 1   city       985 non-null    object 
 2   zip        985 non-null    int64  
 3   state      985 non-null    object 
 4   beds       985 non-null    int64  
 5   baths      985 non-null    int64  
 6   sq__ft     985 non-null    int64  
 7   type       985 non-null    object 
 8   sale_date  985 non-null    object 
 9   price      985 non-null    int64  
 10  latitude   985 non-null    float64
 11  longitude  985 non-null    float64
dtypes: float64(2), int64(5), object(5)
memory usage: 92.5+ KB


In [10]:
housedata['zip'] = housedata['zip'].astype(str)

In [11]:
housedata.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 985 entries, 0 to 984
Data columns (total 12 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   street     985 non-null    object 
 1   city       985 non-null    object 
 2   zip        985 non-null    object 
 3   state      985 non-null    object 
 4   beds       985 non-null    int64  
 5   baths      985 non-null    int64  
 6   sq__ft     985 non-null    int64  
 7   type       985 non-null    object 
 8   sale_date  985 non-null    object 
 9   price      985 non-null    int64  
 10  latitude   985 non-null    float64
 11  longitude  985 non-null    float64
dtypes: float64(2), int64(4), object(6)
memory usage: 92.5+ KB


In [14]:
housedata['sale_date'] = pd.to_datetime(housedata['sale_date'].str.replace('EDT', 'UTC+4'))

AttributeError: Can only use .str accessor with string values!

In [15]:
housedata.head()

street        city    zip state  beds  baths  sq__ft  \
0      3526 HIGH ST  SACRAMENTO  95838    CA     2      1     836   
1       51 OMAHA CT  SACRAMENTO  95823    CA     3      1    1167   
2    2796 BRANCH ST  SACRAMENTO  95815    CA     2      1     796   
3  2805 JANETTE WAY  SACRAMENTO  95815    CA     2      1     852   
4   6001 MCMAHON DR  SACRAMENTO  95824    CA     2      1     797   

          type                 sale_date  price   latitude   longitude  
0  Residential 2008-05-21 00:00:00-04:00  59222  38.631913 -121.434879  
1  Residential 2008-05-21 00:00:00-04:00  68212  38.478902 -121.431028  
2  Residential 2008-05-21 00:00:00-04:00  68880  38.618305 -121.443839  
3  Residential 2008-05-21 00:00:00-04:00  69307  38.616835 -121.439146  
4  Residential 2008-05-21 00:00:00-04:00  81900  38.519470 -121.435768

In [22]:
housedata.describe()

beds       baths       sq__ft          price    latitude  \
count  985.000000  985.000000   985.000000     985.000000  985.000000   
mean     2.911675    1.776650  1314.916751  234144.263959   38.607732   
std      1.307932    0.895371   853.048243  138365.839085    0.145433   
min      0.000000    0.000000     0.000000    1551.000000   38.241514   
25%      2.000000    1.000000   952.000000  145000.000000   38.482717   
50%      3.000000    2.000000  1304.000000  213750.000000   38.626582   
75%      4.000000    2.000000  1718.000000  300000.000000   38.695589   
max      8.000000    5.000000  5822.000000  884790.000000   39.020808   

        longitude  
count  985.000000  
mean  -121.355982  
std      0.138278  
min   -121.551704  
25%   -121.446127  
50%   -121.376220  
75%   -121.295778  
max   -120.597599

In [27]:
housedata[housedata['baths'] == 0]

street             city    zip state  beds  baths  \
73              17 SERASPI CT       SACRAMENTO  95834    CA     0      0   
89           2866 KARITSA AVE       SACRAMENTO  95833    CA     0      0   
100     12209 CONSERVANCY WAY   RANCHO CORDOVA  95742    CA     0      0   
121       5337 DUSTY ROSE WAY   RANCHO CORDOVA  95742    CA     0      0   
126       2115 SMOKESTACK WAY       SACRAMENTO  95833    CA     0      0   
..                        ...              ...    ...   ...   ...    ...   
901  1530 TOPANGA LN Unit 204          LINCOLN  95648    CA     0      0   
917            501 POPLAR AVE  WEST SACRAMENTO  95691    CA     0      0   
934  1550 TOPANGA LN Unit 207          LINCOLN  95648    CA     0      0   
947     1525 PENNSYLVANIA AVE  WEST SACRAMENTO  95691    CA     0      0   
970             3557 SODA WAY       SACRAMENTO  95834    CA     0      0   

     sq__ft         type                 sale_date   price   latitude  \
73        0  Residential 2008-05-21 00:00:00-04:00  206000  38.631481   
89        0  Residential 2008-05-21 00:00:00-04:00  244500  38.626671   
100       0  Residential 2008-05-21 00:00:00-04:00  263500  38.553867   
121       0  Residential 2008-05-21 00:00:00-04:00  320000  38.528575   
126       0  Residential 2008-05-21 00:00:00-04:00  339500  38.602416   
..      ...          ...                       ...     ...        ...   
901       0        Condo 2008-05-15 00:00:00-04:00  138000  38.884150   
917       0  Residential 2008-05-15 00:00:00-04:00  165000  38.584526   
934       0        Condo 2008-05-15 00:00:00-04:00  188000  38.884170   
947       0  Residential 2008-05-15 00:00:00-04:00  200100  38.569943   
970       0  Residential 2008-05-15 00:00:00-04:00  224000  38.631026   

      longitude  
73  -121.501880  
89  -121.525970  
100 -121.219141  
121 -121.228600  
126 -121.542965  
..          ...  
901 -121.270277  
917 -121.534609  
934 -121.270222  
947 -121.527539  
970 -121.501879  

[108 rows x 12 columns]

### 2. Predict Price

We are going to predict `price` from available information.

#### 2.1 What is the Target Variable? Why?

In [29]:
# ANSWER

In [ ]:
#Price, because this is the value that we want to predict

#### 2.2 List all possible variables which might be Independent/Predictor variable.

In [30]:
# ANSWER

In [ ]:
#Beds, Baths, Sq_ft

#### 2.3 Find correlation between variables

Find correlation between variables. Which is the best predictor? Why? State your reason.

In [32]:
# ANSWER
housedata.corr()

beds     baths    sq__ft     price  latitude  longitude
beds       1.000000  0.839650  0.684041  0.342024 -0.291307  -0.053993
baths      0.839650  1.000000  0.665690  0.416208 -0.235792   0.016582
sq__ft     0.684041  0.665690  1.000000  0.333897 -0.383274  -0.177742
price      0.342024  0.416208  0.333897  1.000000 -0.039573   0.284485
latitude  -0.291307 -0.235792 -0.383274 -0.039573  1.000000   0.387925
longitude -0.053993  0.016582 -0.177742  0.284485  0.387925   1.000000

#### 2.4 Find Coefficient and Intercept using Linear Regression

Using LinearRegression of sklearn.linear_model package find Coefficient and Intercept.

Create an instance of LinearRegression.

Explore following methods:

- fit
- predict
- score

In [18]:
from sklearn.linear_model import LinearRegression

##### 2.4.1 Fit predictor and target variables using linear regression 

In [19]:
# ANSWER

X = housedata['sq__ft'].values.reshape(-1,1)
Y = housedata['price']

linreg = LinearRegression()

In [20]:
X

array([[ 836],
       [1167],
       [ 796],
       [ 852],
       [ 797],
       [1122],
       [1104],
       [1177],
       [ 941],
       [1146],
       [ 909],
       [1289],
       [ 871],
       [1020],
       [1022],
       [1134],
       [ 844],
       [ 795],
       [ 588],
       [1356],
       [1118],
       [1329],
       [1240],
       [1601],
       [ 901],
       [1088],
       [ 963],
       [1119],
       [1380],
       [1248],
       [1039],
       [1152],
       [1380],
       [1116],
       [1039],
       [1418],
       [1082],
       [1472],
       [1146],
       [ 760],
       [1304],
       [1207],
       [1056],
       [1043],
       [1587],
       [1120],
       [1580],
       [1955],
       [1656],
       [1477],
       [1188],
       [1590],
       [1463],
       [1714],
       [1185],
       [1406],
       [1943],
       [1172],
       [1152],
       [1851],
       [1215],
       [1130],
       [1603],
       [1479],
       [1420],
       [1280],
       [15

In [21]:
Y

0       59222
1       68212
2       68880
3       69307
4       81900
        ...  
980    232425
981    234000
982    235000
983    235301
984    235738
Name: price, Length: 985, dtype: int64

In [22]:
linreg.fit(X,Y)

LinearRegression()

In [23]:
linreg.coef_

array([54.15863969])

In [24]:
linreg.intercept_

162930.16140699777

##### 2.4.2 Find R Score

Find $R^2$ Using ```score``` method of Linear Regression.

In [25]:
# ANSWER
linreg.score(X,Y)

0.11148717682951437

### 3. Splitting Data

In [26]:
from sklearn.model_selection import train_test_split

#### 3.1 Create training and testing subsets

Hint: Using `train_test_split` Library.

```
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)
```

In [27]:
# ANSWER
## Create training and testing subsets

X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size = 0.2, random_state = 7)

#### 3.2 Check Shape, Sample of Test Train Data

In [28]:
# ANSWER
## Check training/test data
print(X_train.shape)
print(Y_train.shape)
print(Y_test.shape)
print(X_test.shape)

(788, 1)
(788,)
(197,)
(197, 1)


#### 3.3 Using Linear Regression Find The Score

1. Fit model using X_train, y_train
2. Find score using X_test, y_test

In [29]:
# ANSWER
linreg.fit(X_train, Y_train)


LinearRegression()

In [108]:
# ANSWER
print(linreg.score(X_train, Y_train))
print(linreg.coef_, linreg.intercept_)

0.13316227821374926
[60.10388034] 156247.57769344922


In [109]:
print(linreg.score(X_test, Y_test))
print(linreg.coef_, linreg.intercept_)

0.003767424048432666
[60.10388034] 156247.57769344922


### 4. Predict House Price

Let's assume we have information of following house:

- street:	1140 EDMONTON DR
- city:	SACRAMENTO
- zip:	95833
- state:	CA
- beds:	3
- baths:	2
- sq__ft:	1204
- type:	Residential

**Predict the price of this house using linear regression model.**

In [65]:
# ANSWER

In [110]:
Sq_ftsample = 1204
m = linreg.coef_
yo = linreg.intercept_

SamplePrice = yo + m*Sq_ftsample
print(SamplePrice)

[228612.64962404]


#### Find the error

In [80]:
# Predict price from X_test

In [111]:
x = [[1204]]
ypred = linreg.predict(x)
print(ypred[0])

228612.64962403913


In [112]:
y_preds = linreg.predict(X_test)

In [113]:
from sklearn.metrics import mean_squared_error

In [114]:
MSE  = mean_squared_error(Y_test, y_preds, squared = True)

In [115]:
np.sqrt(MSE)

130488.94383342237



---



---



> > > > > > > > > © 2022 Institute of Data


---



---



